### PR TITLE
Optimize and add multi-threading to occlude_image_batch()

### PR DESCRIPTION
**Pull request by schencej**

### CODE REVIEW

**Review by Purg — COMMENTED** [review]
The benchmarking function should have additions to include variations of calling the batch version with threads. I've done this locally and can contribute that as a commit here.

**Comment by Purg** [comment]
> I didn't dive too deeply into the code, but I'm a little confused how this is different from the `occlude_image_streaming` approach? Both seem to parallelize matrix multiplication with the image over a set of perturbation masks.

Given this update, the difference is now in the interface to the method: streaming version is different in that the input masks are an unbounded iterable and the output is a generator for yielding results. I think it is the case still that the streaming version uses less memory (unless the user aggregates the masks, which is usually the case).

Below is a paste of a call to an updated version of the benchmarking function with a larger image size and more thread options my laptop can handle:
```python
In [7]: benchmark_occlude_image(img_shape=(1080, 1920), threading_tests=(1, 2, 4, 8, 16))
Image shape=(1080, 1920, 3), masks=(1000, 1080, 1920), fill_1c=0, fill_3c=[0, 0, 0]
Batch - no-fill - 9.473836011000003 s
Batch - threads= 1 - no-fill - 8.324400905999937 s
Batch - threads= 2 - no-fill - 4.348005320000084 s
Batch - threads= 4 - no-fill - 2.444392760000028 s
Batch - threads= 8 - no-fill - 1.7430356879999636 s
Batch - threads=16 - no-fill - 1.7794073189998016 s
Streaming - threads= 1 - no-fill - 8.902592844999617 s
Streaming - threads= 2 - no-fill - 4.869777243000044 s
Streaming - threads= 4 - no-fill - 2.9124801119996846 s
Streaming - threads= 8 - no-fill - 1.9481475560000945 s
Streaming - threads=16 - no-fill - 1.9174065719998907 s
Batch - fill-1c - 18.20188678500017 s
Batch - threads= 1 - fill-1c - 17.07212517800008 s
Batch - threads= 2 - fill-1c - 8.855388753999705 s
Batch - threads= 4 - fill-1c - 4.75330757599977 s
Batch - threads= 8 - fill-1c - 2.8915296119998857 s
Batch - threads=16 - fill-1c - 2.8811109650000617 s
Streaming - threads= 1 - fill-1c - 17.6275299859999 s
Streaming - threads= 2 - fill-1c - 9.472642456999893 s
Streaming - threads= 4 - fill-1c - 5.363946207000026 s
Streaming - threads= 8 - fill-1c - 3.4406558290002067 s
Streaming - threads=16 - fill-1c - 3.195596307999949 s
Batch - fill-3c - 26.598510732999785 s
Batch - threads= 1 - fill-3c - 25.012394723999932 s
Batch - threads= 2 - fill-3c - 12.77465600000005 s
Batch - threads= 4 - fill-3c - 6.813986526000008 s
Batch - threads= 8 - fill-3c - 4.109469875999821 s
Batch - threads=16 - fill-3c - 4.305397938999704 s
Streaming - threads= 1 - fill-3c - 34.830219158999626 s
Streaming - threads= 2 - fill-3c - 18.047976915000163 s
Streaming - threads= 4 - fill-3c - 10.232103422999899 s
Streaming - threads= 8 - fill-3c - 7.4493127509999795 s
Streaming - threads=16 - fill-3c - 7.527172571999927 s

```

**Comment by schencej** [comment]
> I've done this locally and can contribute that as a commit here.

Please do.

File: xaitk_saliency/utils/masking.py [hunk 1]
```diff
@@ -60,6 +61,9 @@ def occlude_image_batch(
     :param fill: Optional fill for alpha-blending based on the input masks for
         the occluded regions as a scalar value, a per-channel sequence or a
         shape-matched image.
+    :param threads: Optional number of threads to use for parallelism when set
+        to a positive integer. If 0, a negative value, or `None`, work will be
+        performed on the main-thread in-line.
```
[the remark this answers is not in the document]
Comment (schencej) [reply xaitk_saliency/utils/masking.py:66]: Does this advice apply to `occlude_image_streaming` as well? I copied this description directly from there.

[the remark this answers is not in the document]
Comment (Purg) [reply xaitk_saliency/utils/masking.py:66]: Yeah, "advice" would probably be similar/same for both cases. I don't have any specific suggestions, so let's just leave this as is for now until we have a better issue-case.

**Comment by brianhhu** [comment]
Ok, thanks @Purg for putting up the performance test numbers. Is there an easy way to benchmark memory usage alongside speed?

Do we want to expose the `threads` parameter of this new `occlude_image_batch` function (I'm thinking specifically in the `gen_coco_sal` utility function)? It looks like we already do this for `occlude_image_streaming` when using the `PerturbationOcclusion` interface.

**Review by brianhhu — COMMENTED** [review]
Overall, this LGTM- thanks @schencej for adding in this optimization. I added a comment about potentially benchmarking memory usage if possible (that could help distinguish memory-constrained cases where batch vs. streaming may be more appropriate). Also, threading of the `occlude_image_batch` function is not currently being utilized elsewhere in our codebase, which may not be an issue.

**Comment by Purg** [comment]
Bench-marking memory is not specifically trivial and from what I understand include using external tools and plots.

Not fully utilizing the threading option everywhere is OK. We can add it as appropriate.

**Review by Purg — APPROVED** [review]

**Pull request merged**